src/acl_kernel_if.cpp

```diff
@@ -400,20 +400,42 @@ static uintptr_t acl_kernel_cra_set_segment_rom(acl_kernel_if *kern,
 
 static int acl_kernel_cra_read(acl_kernel_if *kern, unsigned int accel_id,
                                unsigned int addr, unsigned int *val) {
+  int result = 0;
   assert(kern->cra_ring_root_exist);
-  uintptr_t segment_offset = acl_kernel_cra_set_segment(kern, accel_id, addr);
-  acl_assert_locked_or_sig();
-  return acl_kernel_if_read_32b(
-      kern, (unsigned)OFFSET_KERNEL_CRA + (unsigned)segment_offset, val);
+#ifdef __linux__
+  acl_sig_block_signals();
+#endif
+  {
+    std::lock_guard<std::mutex> lock(kern->segment_mutex);
+    uintptr_t segment_offset = acl_kernel_cra_set_segment(kern, accel_id, addr);
+    acl_assert_locked_or_sig();
+    result = acl_kernel_if_read_32b(
+        kern, (unsigned)OFFSET_KERNEL_CRA + (unsigned)segment_offset, val);
+  }
+#ifdef __linux__
```

Comment:
Just a general comment.  It would be preferable to have a class whose constructor blocks the signals and whose dtor unblocks them.  Then you don't have to add expliciit code before every return (and would also handle exceptions.
You can make the ctor/dtor only actually do something on Linux, and then you can add it without an #ifdef to a place like this

You also wouldn't need to add a dummy result

Reply:
Okay I created a class for linux signal blocking, I guess I could potentially pass the mutex to the ctor and dtor for locking/unlocking, but I'm not sure if this is making the blocker a little bit "too smart" since there are other place (acl_hal_mmd) where the blockers are used without the mutex?